### ISSUE TITLE
StreamWrapper re-implementation

### ISSUE THREAD

**Issue by blazekadam** [Cognexa/cxflow#166]
Current implementation of `StreamWrapper` is problematic with `buffer`>0. Up to date the following issues were encountered:

- T train increases as T read data decreases (which questions if there is any gain from the buffering)
- stopping and resuming an epoch with cxstream dataset sometimes causes segaults (minimal working example is not available, the problem disappeared with `buffer=0`)

It is suggested to re-implement `StreamWrapper` with either `asyncio` or python `concurrent` futures.

**Comment by petrbel** [comment]
@Teyras afaik mentioned gevent?

**Comment by Teyras** [comment]
I have no idea what are you guys talking about, but asyncio and gevent are both suited for IO-bound tasks. For CPU-bound tasks, you're better off using concurrent.

**Comment by blazekadam** [comment]
I finally had some time to investigate this issue. Apparently, `StreamWrapper` works as expected; nonetheless, the time measurements are confusing. 

In short, I was able to save roughly 70% of epoch wall-clock time which could be potentially saved.

In particular, I carried out an experiment with one of our projects which utilizes **cxtream**. The following table captures [T Read] / [T Train] ([Epoch wall-clock]) values for different buffer set ups.

|   | cxflow no buffer | cxflow buffer=4  |
|---------------------------------|------------------|------------------|
| **cxtream no buffer**               | 46/23 (69)       | 1.6/52 (53)      |
| **cxtream buffer = 4**             | 0.6/23 (24)      | 0.03 / 24.5 (25) |

We can see that without **cxtream** buffer, TRead is greater than TTrain, hence we should be able to save some time with GPU-CPU parallelism, 23 seconds in particular. Note that this saving is only theoretical as TTrain includes the whole `mode.run` method call rather than pure GPU time.

Apparently, **cxflow** buffer indeed saves fair amount of time. More precisely, 16 seconds per epoch which is roughly 70% of the theoretical saving. Not bad huh?

Now the counter-intuitive thing here is that TTrain was larger despite wall clock time being smaller. With some logging in both `StreamWrapper` and `Model.run` method, I discovered that with **cxflow** buffer,  `model.run` method is being interrupted at different places for computing new batches. Those interruptions are of course included in TTrain.

Given all that, it is questionable whether we want to change the implementation and if so, how? @FloopCZ  @petrbel any comments?

**Comment by FloopCZ** [comment]
Thank you for digging into this. I believe that when we saw the issue for the first time, the TRead really decreased, but TTrain increased by more than the decrease in TRead. It is possible that we were not looking at it right though (i believe this was the case in the `isletnet` project). Furthermore, the stream wrapper did not seem to be buffered across the end of the underlying stream with `fixed_epoch_size` option turned on.

Your measurement seems to be solid and the current state is not that bad, so I agree with the removal of the `critical` label.

Also, you say that `model.run` is being interrupted at different places. This is a good observation (well done!), but it is probably not the desired behavior. The buffer interrupts should ideally happen only while the GIL is released by Tensorflow (that was actually the original idea, right?), otherwise we waste a lot of time, because the execution of the graph itself may have to wait until the next batch is prepared. I can imagine the buffer to be idle until it is informed that the control has been moved to Tensorflow's `feed` method, but this would require some new functions like `enable_buffer()` and `disable_buffer()` which could be called from the `Model` and the user could decide when to allow buffering and risk being interrupted (AFAIK there is no way to find out whether a thread is holding the GIL from the inside of Python, unless you use some C code).

**Comment by blazekadam** [comment]
> TRead really decreased, but TTrain increased by more than the decrease in TRead

Same happened in my experiments. In the second row (with **cxtream** buffer), epoch wall clock time increased with **cxflow** buffer turned on. I suppose its because only very little time can be saved in this case (0.6s of TRead) while threading surely introduces some overhead.

In other words, the heavy lifting is done by **cxtream** and **cxflow** can hardly optimize it any further.

It seems that one step further would be to use processing instead of threading. Manual threading management as you suggest would be quite complicated imo...

**Comment by FloopCZ** [comment]
> It seems that one step further would be to use processing instead of threading. Manual threading management as you suggest would be quite complicated imo...

Multiprocessing is something I would rather avoid, though, the overhead is even more significant. :-/
The manual threading management would not be difficult to implement (basically `enable_buffer()` would release some [Lock](https://docs.python.org/3/library/threading.html#threading.Lock) and `disable_buffer()` would aquire the lock; the buffer itself would need an acquired lock to process the next batch). The issue is, that the person who implements the `Model` would have to check whether the function he uses releases the GIL (e.g., PIL does not release the GIL, OpenCV does) and if so, he would have to wrap the function call in the buffer functions. Or we could create a context object, which may be used as:
```python
# those functions need GIL, do not interrupt them
prepare_session()
prepare_towers()
prepare_feed_data()
...
# those functions release GIL, prepare the next batch in the meantime
with BufferEnabled:
    do_crazy_opencv_stuff()
    feed_tensorflow_graph()
```

**Comment by blazekadam** [comment]
Any particular reason why to avoid processing?

Why work around GIL if we can avoid it? We can (and we should) process new batches even during `prepare_session`, `prepare_towers` etc...

**Comment by FloopCZ** [comment]
The memory footprint of multiprocessing is much higher, the communication costs are much higher (pickle) and I am not really sure how is it going to play out with tensorflow, opencv, etc.

I don't believe we should process new batches during `prepare_session`, `prepare_towers`, and other functions holding the GIL, because I consider the **cxflow**'s buffer to be a way to maximally utilize the main Python process by using the GIL for useful work whenever it is released. Not as a tool for efficient parallel processing, that is kind of impossible to do right in Python. For the purpose of efficient parallel processing, you would have to create a multiprocessing pool and be ready for large overheads.

**Comment by petrbel** [comment]
> The memory footprint of multiprocessing is much higher, the communication costs are much higher (pickle) and I am not really sure how is it going to play out with tensorflow, opencv, etc.

The same problem we fought in old fuel streams... So I'd avoid multiprocessing as well.

I remember @FloopCZ mentioned [concurrent.futures](https://docs.python.org/3/library/concurrent.futures.html#module-concurrent.futures) module which should easily solve all out problems with a simple thread pool. Is there any problem with using it? Futhermore, the buffer-related code would be significantly more readable.

**Comment by FloopCZ** [comment]
`concurrent` is a nice tool and yes, it would probably make the code cleaner. Unfortunately, it would not solve all our problems. The main problem is that that when a new batch is being processed in the buffer, the buffer holds the GIL. That is ok, but it should happen only when we have nothing else to do (e.g., in TF's `feed` method), not all the time (see https://github.com/Cognexa/cxflow/issues/166#issuecomment-341962285), because otherwise, the graph execution might actually have to wait for buffer even though there are some batches already prepared. It could be solved by:

- Manually informing the buffer to run when I am sure the GIL will be released (see https://github.com/Cognexa/cxflow/issues/166#issuecomment-341968096).
- Multiprocessing. This solution is more out-of-the-box as there would be no need to communicate with the buffer in the `Model`. On the other hand, it is less efficient and in some moment, you have to read the prepared batch from a different process (i.e., `unpickle`) anyway, which is again a GIL operation. I am not sure whether serialization and deserialization of e.g., a large numpy array is fast enough and if the overhead won't actually make it slower.

Now the question is, what is the right way to go. I have not decided, but I am more inclined to the first option, since it is easy to implement (even with the current buffer implementation) and more efficient. I would rather let the heavy multithreading to be implemented in the stream.

**Comment by blazekadam** [comment]
Well, fuel socket producer-consumer worked pretty well imo. The issues were caused by theano GPU management and parallel batch processing...

Anyways, how important is the re-implementation in your opinion?

**Comment by FloopCZ** [comment]
Not that much, there are probably more important things to do and it won't affect the API that horribly. Do you agree with tagging for release 0.12 ?

**Comment by petrbel** [comment]
> Anyways, how important is the re-implementation in your opinion?

Personally, not that much.

However, from the PR point of view, it's a problem since `LogProfile` hook reports invalid values (even though they are correct in total as you investigated).

I suggest keeping this issue open (or creating a brand new one if you want) and eventually solve this. For me, the question is whether to do so right now (imho necessary).

**Comment by petrbel** [comment]
We tried to measure and compare the T read during training (via cxflow) and just by iterating through the stream (in python). The score is 43s vs. 19s which is kinda alarming.

I think the @FloopCZ's solution with yielding the threads + a single mutex is very easy to implement. I vote for it.

Btw see `time.sleep(0)`, it will be useful.

**Comment by FloopCZ** [comment]
Thank you for the profiling @petrbel, this needs to be fixed.

**Comment by FloopCZ** [comment]
I have implemented a dirty prototype in `stream-wrapper` branch of `cxflow` and `cxflow-tensorflow`.

It is necessary to sleep at least 0.02s (0.002s does not work) in the worker after acquiring the lock. Then the data reading (i.e., the 45s) gets almost completely hidden under the training time (only 3s remain). I will finalize the change into a PR in the next few days.

**Comment by petrbel** [comment]
> It is necessary to sleep at least 0.02s (0.002s does not work)

Really? I have a feeling that you mentioned that even `time.sleep(0)` would yield. Will 0.02 work on other machines? Should the timeout be configurable?

**Comment by blazekadam** [comment]
Removing the **critical** label as it is more or less addressed with #199. I would close it actually. @FloopCZ ?

**Comment by FloopCZ** [comment]
I agree.
